File: .claude/openspec-validate-on-edit.py
#!/usr/bin/env python3
import json, re, subprocess, sys

data = json.load(sys.stdin)
path = data.get("tool_input", {}).get("file_path", "")

match = re.search(r"openspec/changes/([^/]+)/", path)
if not match or match.group(1) == "archive":
    sys.exit(0)

change_name = match.group(1)
result = subprocess.run(
    ["bunx", "openspec", "validate", change_name, "--strict"],
    capture_output=True,
    text=True,
)

if result.returncode != 0:
    sys.stderr.write(f"openspec validate {change_name} --strict failed:\n{result.stdout}{result.stderr}")
    sys.exit(2)

File: .agents/hooks/openspec-validate-on-edit.ts
#!/usr/bin/env bun

import { readHookInput } from "./claude-hook";

const CHANGE_PATH_PATTERN = /openspec\/changes\/([^/]+)\//;

const input = await readHookInput();
const path = input.tool_input?.file_path ?? "";

const match = path.match(CHANGE_PATH_PATTERN);
if (!match || match[1] === "archive") {
  process.exit(0);
}

const changeName = match[1];
const proc = Bun.spawn(["bunx", "openspec", "validate", changeName, "--strict"], {
  stdout: "pipe",
  stderr: "pipe",
});
const [stdout, stderr, exitCode] = await Promise.all([
  new Response(proc.stdout).text(),
  new Response(proc.stderr).text(),
  proc.exited,
]);
if (exitCode !== 0) {
  process.stderr.write(
    `openspec validate ${changeName} --strict failed:\n${stdout}${stderr}`,
  );
  process.exit(2);
}
